Implements EventSubscriber in GithubOrgEntityProvider

angeliski opened the pull request:
…ership  events

Signed-off-by: Rogerio Angeliski <angeliski@hotmail.com>

## Hey, I just made a Pull Request!

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

angeliski commented on plugins/catalog-backend-module-github/src/providers/GithubOrgEntityProvider.ts:
```diff
@@ -224,6 +243,202 @@ export class GithubOrgEntityProvider implements EntityProvider {
+
+    await this.connection.refresh({
+      keys: [
+        `url:https://${this.options.gitHubConfig.host}/orgs/${org}/teams/${teamSlug}`,
+        `url:https://${this.options.gitHubConfig.host}/${login}`,
```
I put this refresh here, but I don't know why the group and the user stay without the refresh. Is the key wrong? 

@kissmikijr could you help me here please? I am not seeing what I am doing wrong

kissmikijr replied:
Unfortunately it won't work for the current implementation of the org entity provider. The connection.refresh() works when there is a matching refresh_key for the entity. Currently this refresh_key is added by the UrlReaderProcessor to the Location entities. In the case of this provider it doesn't use Locations for the individual teams/users. The way it works is that it directly provides the raw entities, not the Locations for them.

I think here we will need to rely on the catalogApi and handle the event by either updating an already existing group/user or creating new ones when the event comes in.

angeliski replied:
Thanks, I will change to use catalogApi in this case

angeliski commented:
I am having some problems to handle the change events.
For the [team.edited](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#team) I can't correlate the grouo entity when the name changes, github don't send me the "old slug" neither can I trust the transformer will only use this for the entity.

For the [membership](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#membership) I need to create/remove relations, but I can't emit relations here.

I could find those entities in the catalog and use a delta mutation to remove and add the new data, but if something goes wrong between those operations the state will stay wrong until the next cicle, 

so I will avoid those events for now, until I got a better ideia how to handle those problems

kissmikijr commented:
> For the [team.edited](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#team) I can't correlate the grouo entity when the name changes, github don't send me the "old slug" neither can I trust the transformer will only use this for the entity.

In the linked docs they say they sand the old name of the group in the `body.changes.name.from` or is this not enough for us ?

>I could find those entities in the catalog and use a delta mutation to remove and add the new data, but if something goes wrong between those operations the state will stay wrong until the next cicle,

Just to be clear in between which operations? If you emit a mutation from an entity provider it will be put in the front of the queue.
I think this approach will be good enough, if it is finally able to resolve itself by a full scan (for example) this should be fine.

angeliski commented:
> In the linked docs they say they sand the old name of the group in the body.changes.name.from or is this not enough for us ?

No, because this is the old team name, not the slug. And when you change the name, the slug changes too 😢 

Example:
I changed the name from  `my-group-name` to `My new Group`. This will create a new slug, but I don't receive the old one to search in the catalog.

```json
{
  "changes": {
    "name": {
      "from": "My group name"
    }
  },
  "action": "edited",
  "team": {
    "name": "My new Group",
    "id": 7058073,
    "node_id": "T_kwDOBCL6C84Aa7KZ",
    "slug": "my-new-group",
    "description": "",
    "privacy": "closed",
    "url": "https://api.github.com/organizations/69401099/team/7058073",
    "html_url": "https://github.com/orgs/rangeliski/teams/my-new-group",
    "members_url": "https://api.github.com/organizations/69401099/team/7058073/members{/member}",
    "repositories_url": "https://api.github.com/organizations/69401099/team/7058073/repos",
    "permission": "pull",
    "parent": null
  }
}
```

Another problem is: I can't use the `refresh` from catalogApi, because we don't have a processor for this, so I need to really to update this entity when receive the event (using a mutation).


> Just to be clear in between which operations? If you emit a mutation from an entity provider it will be put in the front of the queue.
I think this approach will be good enough, if it is finally able to resolve itself by a full scan (for example) this should be fine.

I was thinking something like this:
```typescript
    const team = await catalogClient.getEntityByRef(
      { kind: 'Group', namespace:  'default', name: slug },
      { token },
    );

addNewRelation(team, event)

// remove group from catalog
await this.connection.applyMutation({
      type: 'delta',
      removed:[team],
      added:[],
    });

// if something goes wrong HERE, we will get a dirty state

// add group update in the catalog
await this.connection.applyMutation({
      type: 'delta',
      removed:[],
      added:[team],
    });
```

kissmikijr commented:
> I changed the name from My group name to My new Group. This will create a new slug, but I don't receive the old one to search in the catalog.

I think we have a couple of options. 
1. We could go and construct the slug from the name, as far as I can see it looks like they switch uppercases to lower cases and swaps out whitespaces with dashes.
2. We could update the existing provider to start to store the original group name somewhere on the entity as an annotation and we could rely on it in this.

> addNewRelation(team, event)

Here I don't think we need to add a relation, we just need to find the group entity, and update its `spec.members` array and apply the mutation with the new user entity and the processing loop will take care of creating the relations


AFAIK you don't need to remove the old team entity by hand, if you perform a applyMutation with an added array it is actually add and update for the entities, if there is already an entity with that entity ref.

angeliski commented:
> We could update the existing provider to start to store the original group name somewhere on the entity as an annotation and we could rely on it in this.

Nice, I will try to store the id from github, maybe this will avoid some other data to store
> Here I don't think we need to add a relation, we just need to find the group entity, and update its spec.members array and apply the mutation with the new user entity and the processing loop will take care of creating the relations
>
> AFAIK you don't need to remove the old team entity by hand, if you perform a applyMutation with an added array it is actually add and update for the entities, if there is already an entity with that entity ref.

🤔 I will try some tests to see the behavior in this case

angeliski commented:
@kissmikijr @pjungermann I think we are ready for a first review

Thanks for the big help @kissmikijr, your suggestions helped me a lot to find a worth solution

jhaals commented on docs/integrations/github/org.md:
```diff
@@ -113,6 +165,10 @@ There is also a `defaultUserTransformer` and `defaultOrganizationTeamTransformer
 
+**Attention:**
+When you use the Events Support with a `TeamTransformer`, you need to ensure your entity will have the
+annotation `github.com/node-id` with the value from `team.id`. Without this, the `team.edit` event could fail to update
```
```suggestion
annotation `github.com/team-id` with the value from `team.id`. Without this, the `team.edit` event could fail to update
```

hmm, wouldn't it be clearer for someone inspecting the entity to find team-id instead of node-id?

angeliski replied:
I thought that, but the thing is: We have an `id` for the team and isn't this information. I can change, but this could create some confusion in the future. I will see if the team id could be returned in the graphql query

angeliski replied:
Updated to use the **real** team id :)

jhaals commented on plugins/catalog-backend-module-github/src/providers/GithubOrgEntityProvider.ts:
```diff
@@ -224,6 +261,329 @@ export class GithubOrgEntityProvider implements EntityProvider {
+    // throw only once
+    if (!this.eventConfigErrorThrown) {
+      this.eventConfigErrorThrown = true;
+      throw new Error(
```
Is it okay to throw here or is log.warn/error really what we want instead? I'm thinking if this causes the rest of the execution to stop.

angeliski replied:
Same discussion from [here](https://github.com/backstage/backstage/pull/13931#discussion_r1001767291).
The best option should throw this error at backend start 🤔

jhaals replied:
Thanks for sharing the past discussion, lets keep it 👍

jhaals reviewed (changes requested):
Awesome to see event based updates coming to users and groups as well! 👏 💯 I've added some suggestions

jhaals commented on plugins/catalog-backend-module-github/api-report.md:
```diff
@@ -253,7 +253,7 @@ export class GithubOrgReaderProcessor implements CatalogProcessor {
 // @public
 export type GithubTeam = {
-  id: string;
+  databaseId: number;
```
I think I would prefer `id` here tbh. Or does `id` mean something else here? Reading `databaseId` make me think it's something in our postgres database.

I'm open to suggestions perhaps @kissmikijr or @pjungermann has any?

angeliski replied:
I used databaseId here because it's the value comes from the [Github Graphql query](https://github.com/backstage/backstage/pull/14870/files#diff-50740c9b7b864c9aad8c13fe8fb48cdfab8e52f91c8c2cd6ee99cea8c10e7e17R190), I don't know a way to change from the transformation, sorry

Rugvip commented on plugins/catalog-backend-module-github/src/providers/GithubOrgEntityProvider.ts:
```diff
@@ -224,6 +261,335 @@ export class GithubOrgEntityProvider implements EntityProvider {
+
+    const { token } = await this.options.tokenManager!.getToken();
+    const org = event.organization.login;
+    let group = (await this.options.catalogApi!.getEntityByRef(
+      {
+        kind: 'Group',
+        namespace: DEFAULT_NAMESPACE,
+        name: teamSlug,
+      },
+      { token },
+    )) as GroupEntity;
```
I'd like to avoid this pattern. There's no guarantee that the shape of the entity is what you expect here, there can be processors that mutate the entities and the catalog model might've been customized. It's also data flowing in the wrong direction, what we're looking for is `github -> providers -> processors -> output entities -> catalog clients`, without any circling back.

Instead of fetching these entities from the catalog, I think it's best to fetch any information that's missing from GitHub instead. So in this case you might fetch the group and all of its members again, or the user, whichever is easiest. Basically, any time you get an event that something has changes, you go back to the source of truth, in this case GitHub, to fill in just enough information that you're able to emit the updated entities.

CC: @pjungermann ^ 👀 

Bit of a separate note here, but I don't think you need to update both the user and group in this case, should be enough to just update one of them and the catalog processing will fill in the relations in both directions.

pjungermann replied:
Also not so sure about that as a general rule. E.g., at the `BitbucketCloudEntityProvider` I used the `catalogApi` as well. Different use case though. I needed it to check what exists at catalog in order to create the delta. The events don't contain enough information to do this without.

angeliski replied:
@Rugvip I was working here and when team changes name, I don't have enough information to delete the old relation.

I can use a simple approach and rebuild the whole org (calling the `read` again), or drop support for this event.

The member change is tricky too, because when I emit the delta for the user I need to emit all `memberOf`, so for this we can rebuild the whole org (calling the `read` again), or drop support for this event.

The only drawback I can see  for rebuild th whole org  is for big orgs. We have a hunge chunk to process.
The rate limit will be used for any request, so we can't avoid that.

I will change to call the `read`, but want to get your opnion here too

Rugvip replied:
@angeliski for the team name change, I think the old name should be available in the event payload right? at `.changes.name.from`. Or is there some other piece of information that's missing?

For member changes, I'd expect it to be possible to just fetch the entire team and its members again, and apply that as an update?

It'd be best if we can avoid full `.read()` calls except in very rare situations.

Rugvip reviewed (changes requested):
Thank you! 👍 

Bit targeted review to discuss the pattern of fetching things from the catalog

angeliski commented:
@Rugvip @jhaals New year, new try LOL

jhaals commented:
> @Rugvip @jhaals New year, new try LOL

🎉 Nice work, I feel 2023 is going to be great! This is a big improvement from the original implementation avoiding the extra id and reading/writing from the catalog. Just some a asserts that needs tweaking but this looks good to me. 

@Rugvip please take another look as well

Rugvip commented on plugins/catalog-backend-module-github/src/lib/github.ts:
```diff
@@ -238,6 +241,159 @@ export async function getOrganizationTeams(
+            hasNextPage
+          }
+          nodes {
+            login
```
Are we able to fetch the rest of the member fields here as well? Same in the query below

angeliski replied:
Done :)

Rugvip reviewed (commented):
Nice! 😁

jhaals reviewed (approved):
LGTM!
